### PR TITLE
Add spec and verifier for quantized dot_general op

### PR DESCRIPTION
**Pull request by doyeonkim0**
This PR adds spec for hybrid quantized op discussed in the [RFC](https://github.com/openxla/stablehlo/pull/1792) and implements verifier accordingly.
Please feel free to take a look. Thank you!

cc @sdasgup3 @loganchien @abhigunj

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -2590,14 +2590,17 @@ planning to address this in
...
+  * (C15) `zero_points(rhs) = 0`.
+  * (C16) If `is_per_axis_quantized(rhs)`, then
     `quantization_dimension(rhs)` not in `rhs_contracting_dimensions`.
+  * If `is_quantized(lhs)`:
+    * (C17) `storage_type(lhs) = storage_type(rhs)`.
+    * (C18) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+    * (C19) If `is_per_tensor_quantized(rhs)`, then
+      `is_per_tensor_quantized(result)`.
+  * If `!is_quantized(lhs)`:
+    * (C20) `element_type(lhs) = expressed_type(rhs) = element_type(result)`.
```
Comment (loganchien) [inline docs/spec.md:2603]: Please add `C20` to the table above (Line 2556-2570)

File: stablehlo/dialect/TypeInference.cpp [hunk 2]
```diff
@@ -3614,6 +3614,46 @@ LogicalResult verifyDotOp(std::optional<Location> location,
...
+                 dyn_cast<quant::UniformQuantizedPerAxisType>(rhsElementType)) {
+    if (llvm::any_of(rhsPerAxisQuantType.getZeroPoints(),
+                     [](int64_t zero_point) { return zero_point != 0; })) {
+      return emitOptionalError(location,
+                               "Zero points of rhs of dot_general should be 0");
+    }
+  }
+
+  // dot_general_c16
+  if (auto rhsPerAxisType =
```
[the remark this answers is not in the document]
Comment (doyeonkim0) [reply stablehlo/dialect/TypeInference.cpp]: Done.

File: docs/spec.md (hunk 1 above)
Comment (doyeonkim0) [reply docs/spec.md:2603]: Thanks for pointing this out! Updated the table for other constraints and for convolution as well.

**Review by loganchien — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -3614,6 +3614,43 @@ LogicalResult verifyDotOp(std::optional<Location> location,
...
+      return emitOptionalError(location,
+                               "Zero points of rhs of dot_general should be 0");
+    }
+
+    // dot_general_c16
+    if (llvm::find(rhsContractingDimensions,
+                   rhsPerAxisQuantType.getQuantizedDimension()) !=
+        rhsContractingDimensions.end()) {
+      return emitOptionalError(location,
+                               "Quantization dimension should not be a "
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:3644]: Quantization dimension of rhs should not be in the contracting... ?

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp [hunk 4]
```diff
@@ -3614,6 +3614,43 @@ LogicalResult verifyDotOp(std::optional<Location> location,
...
+  Type lhsElementType = getElementTypeOrSelf(lhsType);
+  Type rhsElementType = getElementTypeOrSelf(rhsType);
+  Type resultElementType = getElementTypeOrSelf(resultType);
+
+  // dot_general_c15
+  if (auto rhsPerTensorQuantType =
+          dyn_cast<quant::UniformQuantizedType>(rhsElementType)) {
+    if (rhsPerTensorQuantType.getZeroPoint() != 0) {
+      return emitOptionalError(location,
+                               "Zero point of rhs of dot_general should be 0");
```
Comment (abhigunj) [inline stablehlo/dialect/TypeInference.cpp:3629]: omit op name from the error log here, and other places.

**Review by abhigunj — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (doyeonkim0) [reply stablehlo/dialect/TypeInference.cpp]: Done.

File: stablehlo/dialect/TypeInference.cpp [hunk 5]
```diff
@@ -3614,6 +3614,43 @@ LogicalResult verifyDotOp(std::optional<Location> location,
...
+    if (llvm::any_of(rhsPerAxisQuantType.getZeroPoints(),
+                     [](int64_t zero_point) { return zero_point != 0; })) {
+      return emitOptionalError(location,
+                               "Zero points of rhs of dot_general should be 0");
+    }
+
+    // dot_general_c16
+    if (llvm::find(rhsContractingDimensions,
+                   rhsPerAxisQuantType.getQuantizedDimension()) !=
+        rhsContractingDimensions.end()) {
```
[the remark this answers is not in the document]
Comment (doyeonkim0) [reply stablehlo/dialect/TypeInference.cpp]: Done.

File: stablehlo/dialect/TypeInference.cpp (hunk 4 above)
Comment (doyeonkim0) [reply stablehlo/dialect/TypeInference.cpp]: Done.

**Pull request merged**